source/fe/mapping_cartesian.cc

```diff
@@ -385,6 +385,66 @@ MappingCartesian<dim, spacedim>::maybe_update_jacobian_derivatives(
 
 
 
+template <int dim, int spacedim>
+void
+MappingCartesian<dim, spacedim>::maybe_update_volume_elements(
+  const InternalData &data) const
+{
+  if (data.update_each & update_volume_elements)
+    {
+      double volume = data.cell_extents[0];
+      for (unsigned int d = 1; d < dim; ++d)
+        volume *= data.cell_extents[d];
+      data.volume_element = volume;
+    }
```

Comment:
I would like to highlight that this is really misleading. I had to declare this inparameter as `const`, because`InternalData` is `const` when passed to `fill_fe_*values`. But we actually modify `data` here, `InternalData::volume_element` is `mutable`...